patches/react-native-incall-manager+4.2.0.patch

```diff
@@ -0,0 +1,334 @@
+-                                }
+-                            }
+-                        }, seconds * 1000);
++            if (seconds > 0) {
++                mRingtoneCountDownHandler = new Handler();
++                mRingtoneCountDownHandler.postDelayed(() -> {
++                    try {
++                        Log.d(TAG, String.format("mRingtoneCountDownHandler.stopRingtone() timeout after %d seconds", seconds));
++                        stopRingtone();
++                    } catch (Exception e) {
++                        Log.d(TAG, "mRingtoneCountDownHandler.stopRingtone() failed.");
+                     }
+-
+-                    Looper.loop();
+-                } catch(Exception e) {
+-                    wakeLockUtils.releasePartialWakeLock();
+-                    Log.e(TAG, "startRingtone() failed", e);
+-                }
++                }, seconds * 1000);
+             }
+-        };
++        } catch (Exception e) {
++            wakeLockUtils.releasePartialWakeLock();
++            Log.e(TAG, "startRingtone() failed", e);
++            promise.reject("startRingtone() failed", e);
++        }
+ 
+-        thread.start();
++        promise.resolve(null);
+     }
+ 
++
+     @ReactMethod
+-    public void stopRingtone() {
+-        Thread thread = new Thread() {
+-            @Override
+-            public void run() {
+-                try {
+-                    if (mRingtone != null) {
+-                        mRingtone.stopPlay();
+-                        mRingtone = null;
+-                        restoreOriginalAudioSetup();
+-                    }
+-                    if (mRingtoneCountDownHandler != null) {
+-                        mRingtoneCountDownHandler.removeCallbacksAndMessages(null);
+-                        mRingtoneCountDownHandler = null;
+-                    }
+-                } catch (Exception e) {
+-                    Log.d(TAG, "stopRingtone() failed");
+-                }
+-                wakeLockUtils.releasePartialWakeLock();
+-            }
+-        };
++    public void stopRingtone(Promise promise) {
++        try {
++            stopRingtone();
++        } catch (Exception e) {
++            promise.reject("stopRingtone() failed", e);
++        }
++        promise.resolve(null);
++    }
+ 
+-        thread.start();
++    private void stopRingtone() {
++        try {
++            if (mRingtone != null) {
++                mRingtone.stopPlay();
++                mRingtone = null;
++                restoreOriginalAudioSetup();
++            }
++            if (mRingtoneCountDownHandler != null) {
++                mRingtoneCountDownHandler.removeCallbacksAndMessages(null);
++                mRingtoneCountDownHandler = null;
++            }
++        } catch (Exception e) {
++            Log.d(TAG, "stopRingtone() failed");
++            throw e;
++        }
++        wakeLockUtils.releasePartialWakeLock();
```

Comment:
Would you expect to release the lock in case of error?

Reply:
Ahh, nice catch, thank you.